api/agent/agent.go

```diff
@@ -532,7 +504,37 @@ func (s *coldSlot) exec(ctx context.Context, call *call) error {
 		return res.Error()
 	}
 
-	// nil or timed out
+	// This means default IO with no limits (no clamper)
+	clamper, ok := s.stdout.(*protocol.ClampWriter)
+	if !ok {
+		return ctx.Err()
+	}
+	if clamper.E {
+		return models.NewAPIError(http.StatusBadGateway, fmt.Errorf("invalid response from function err: body too large"))
+	}
+
+	errApp := make(chan error, 1)
+
+	// take ownership of this buffer to avoid cold token.Close() to touch it.
+	buf := clamper.W.(*bytes.Buffer)
+	s.stdout = nil
+	// yes, we leak this and retain the buf until client times out. But this timeout is
+	// closely bounded by our own ctx timeout below. Once we return err, we expect gin
+	// to close/finalize client request/connection.
+	go func() {
+		// IMPORTANT: keep the buffer until client is done.
+		defer protocol.BufPool.Put(buf)
+		_, err := io.Copy(call.w, ioutil.NopCloser(buf))
+		errApp <- err
+	}()
```

Comment:
Dispatch and cold have never played well together, the back story on that is that if you start the container and it ignores stdin then it can exit before dispatch is ready to see what's going on or we've handed out the slot to actually receive the error from the container. it sucks. i wish we could thread cold through Dispatch. maybe someone with a bigger brain than I can solve the puzzle there.

all that's to say, this seems really hairy. this appears to add buffering of default format responses before shipping them out. dispatch was a good spot to put this (for response headers crap, mostly) and i'm sad we can't put this there and I don't think we can thread a buffer as the writer in front of all calls due to the headers crap (that has to be done in dispatch on a ResponseWriter). am trying to think but all jolted on cheap burnt conference coffee and need to eat. i'll leave that to marinate on and be back later... in any case, it seems like we should find a unified place to either tuck this and return a `models.APIError` from the `io.Copy` calls or from somewhere else (there are challenges here with pipes, i'm aware).

Reply:
I could try to move this code into Dispatch(), as default protocol. There's already some dead code in there for cold-dispatch (eg. default.go)

Reply:
I looked at this one more time, but trying to move this into Dispatch() doesn't gain us much, the exec path is significantly different for cold.

Also I think we do want buffering of responses in all cases anyway, right?